libdrgn/program.c

```diff
@@ -788,20 +886,162 @@ struct drgn_error *drgn_program_find_prstatus_by_tid(struct drgn_program *prog,
 						     struct nstring *ret)
 {
 	struct drgn_error *err;
-	struct drgn_prstatus_map_iterator it;
-
-	assert(!(prog->flags & DRGN_PROGRAM_IS_LINUX_KERNEL));
-	err = drgn_program_cache_prstatus(prog);
+	struct drgn_thread *thread;
+	err = drgn_program_find_thread(prog, tid, &thread);
 	if (err)
 		return err;
 
-	it = drgn_prstatus_map_search(&prog->prstatus_map, &tid);
-	if (!it.entry) {
+	if (!thread) {
 		ret->str = NULL;
 		ret->len = 0;
 		return NULL;
 	}
-	*ret = it.entry->value;
+	*ret = thread->prstatus;
+	return NULL;
+}
+
+LIBDRGN_PUBLIC struct drgn_error *
+drgn_program_find_thread(struct drgn_program *prog, uint32_t tid,
+			 struct drgn_thread **ret)
+{
+	struct drgn_error *err;
+
+	if (!(prog->flags & DRGN_PROGRAM_IS_LIVE)) {
```

Comment:
Do we need to cache threads for a kernel core dump here? If not, I'd move this to the `} else {` (not kernel) branch below.